isabelle/haskell/MarloweCoreJson.lhs

```diff
@@ -0,0 +1,1394 @@
+
+
+\isamarkupsubsubsection{TransactionShadowing}
+
+\begin{code}
+transactionShadowingExample :: TransactionWarning
+transactionShadowingExample = TransactionShadowing
+                                            (ValueId "example")
+                                            4
+                                            5
+\end{code}
+
+is serialized as
+\perform{prettyEncodeExample transactionShadowingExample}
+
+
+\isamarkupsubsubsection{TransactionAssertionFailed}
+
+\begin{code}
+transactionAssertionFailedExample :: TransactionWarning
+transactionAssertionFailedExample = TransactionAssertionFailed
+\end{code}
+
+is serialized as \eval{encodeExample transactionAssertionFailedExample}
+
+
+\isamarkupsection{IntervalError}
+
+The \emph{IntervalError} type is serialized as an object with a single property (depending on the constructor)
+and in a tuple, the values.
+\begin{code}
+instance ToJSON IntervalError where
+  toJSON (InvalidInterval (s, e)) = object
+    [ ("invalidInterval" .=  toJSON (s, e)) ]
+  toJSON (IntervalInPastError t (s, e)) = object
+    [ ("intervalInPastError" .=  toJSON (t, s, e)) ]
+
+instance FromJSON IntervalError where
+  parseJSON (JSON.Object v) =
+    let
+      parseInvalidInterval = do
+        (s, e) <- v .: "invalidInterval"
+        pure $ InvalidInterval (s , e)
+      parseIntervalInPastError = do
+        (t, s, e) <- v .: "intervalInPastError"
+        pure $ IntervalInPastError t (s, e)
+    in
+      parseIntervalInPastError <|> parseInvalidInterval
+  parseJSON invalid =
+      JSON.prependFailure "parsing IntervalError failed, " (JSON.typeMismatch "Object" invalid)
+\end{code}
+
+Some examples for each \emph{IntervalError} type
+
+
+\isamarkupsubsubsection{InvalidInterval}
+
+\begin{code}
+invalidIntervalExample :: IntervalError
+invalidIntervalExample = InvalidInterval (10, 20)
+\end{code}
+
+is serialized as \eval{encodeExample invalidIntervalExample}
+
+
+\isamarkupsubsubsection{IntervalInPastError}
+
+\begin{code}
+intervalInPastErrorExample :: IntervalError
+intervalInPastErrorExample = IntervalInPastError 30 (10, 20)
+\end{code}
+
+is serialized as \eval{encodeExample intervalInPastErrorExample}
+
+\isamarkupsection{TransactionError}
+
+The \emph{TransactionError} type is serialized as an object with a \emph{tag} property that differentiates
+the type, and a \emph{contents} property that includes the parameter if any.
+
+\begin{code}
+instance ToJSON TransactionError where
+  toJSON TEAmbiguousTimeIntervalError = object
+    [ "tag" .= JSON.String "TEAmbiguousTimeIntervalError"
+    , "contents" .= JSON.Null
+    ]
+  toJSON TEApplyNoMatchError = object
+    [ "tag" .= JSON.String "TEApplyNoMatchError"
+    , "contents" .= JSON.Null
+    ]
+  toJSON (TEIntervalError e) = object
+    [ "tag" .= JSON.String "TEIntervalError"
+    , "contents" .= toJSON e
+    ]
+  toJSON TEUselessTransaction = object
+    [ "tag" .= JSON.String "TEUselessTransaction"
+    , "contents" .= JSON.Null
+    ]
```

Comment:
Why not call it `error` instead of `tag`?

Reply:
This is how haskell generates automatic codecs when deriving the implementation. I mentioned in the Readme as one of the current specification problems.

Following how we model other types, it should be something more like

`"TEAmbiguousTimeIntervalError"`
| `"TEApplyNoMatchError"`
| `{ "interval_error": ...}`
| `"TEUselessTransaction"`